File: levelGEN.py
# Внизу будет указано, где можно вставить
# название файла, чтобы (поиграться) протестить алгоритм.

from PIL import Image, ImageDraw, ImageColor
from random import *
import math

def lowerImage(img):
    w = img.size[0]
    h = img.size[1]
    w_ = w
    h_ = h
    div = 2
    while w_ > 512 and h_ > 512:
        w_ = w_ // 2
        h_ = h_ // 2
        div *= 2
    img = img.resize((w//div,h//div), Image.ANTIALIAS)
    return img

def getRGB(image, x, y): # очередная бесполезная функция get, пусть будет
    pix = image.load()
    r = pix[x, y][0]
    g = pix[x, y][1]
    b = pix[x, y][2]
    return r, g, b

def rgb2H(r, g, b): # преобр. RGB в параметр тона
    r, g, b = r/255, g/255, b/255
    cmax = max(r, g, b)
    cmin = min(r, g, b)
    diff = cmax-cmin
  
    if cmax == cmin:  
        h = 0
    elif cmax == r:  
        h = (60 * ((g - b) / diff) + 360) % 360
    elif cmax == g: 
        h = (60 * ((b - r) / diff) + 120) % 360
    elif cmax == b: 
        h = (60 * ((r - g) / diff) + 240) % 360
    return round(h)

def rgb2S(r, g, b): # преобр. RGB в параметр насыщенности
    r, g, b = r/255, g/255, b/255
    cmax = max(r, g, b)
    cmin = min(r, g, b)
    diff = cmax-cmin
    if cmax == 0: 
        s = 0
    else: 
        s = (diff / cmax) * 100
    return round(s)

def rgb2V(r, g, b): # преобр. RGB в параметр яркости
    r, g, b = r/255, g/255, b/255
    cmax = max(r, g, b)
    v = cmax * 100
    return round(v)

def hsv2RGB(h, s, v): # преобразование HSV -> RGB
    hi = math.floor(h/60)%6
    vm = (100 - s)*v/100
    a = (v-vm)*(h%60)/60
    vi = vm + a
    vd = v - a
    if hi==0:
        r = v
        g = vi
        b = vm
    elif hi==1:
        r = vd
        g = v
        b = vm
    elif hi==2:
        r = vm
        g = v
        b = vi
    elif hi==3:
        r = vm
        g = vd
        b = v
    elif hi==4:
        r = vi
        g = vm
        b = v
    else:
        r = v
        g = vm
        b = vd
    return round(r*255/100), round(g*255/100), round(b*255/100)

def middleColor(img):
    mid = 0
    r_ = 0
    g_ = 0
    b_ = 0
    pix = img.load()
    for x in range(img.size[0]):
        for y in range(img.size[1]):
            r = pix[x,y][0]
            g = pix[x,y][1]
            b = pix[x,y][2]
            r_ += r
            g_ += g
            b_ += b
            mid += ((r+g+b) / 3)
    r_ /= (img.size[0]*img.size[1])
    g_ /= (img.size[0]*img.size[1])
    b_ /= (img.size[0]*img.size[1])
    mid /= (img.size[0]*img.size[1])
    return round(r_), round(g_), round(b_)
            


def createLayer(img, param): #Создать изображение на основе отдельного канала ('h', 's', 'v')
    width = img.size[0]//2
    height = img.size[1]//2
    new_image = img
    draw = ImageDraw.Draw(new_image)
    pix = new_image.load()
    for x in range(width):
        for y in range(height):
            r = pix[x,y][0]
            g = pix[x,y][1]
            b = pix[x,y][2]
            if param=='h':
                h = rgb2H(r, g, b)
                r, g, b = hsv2RGB(h, 100, 100)
            elif param=='v':
                v = rgb2V(r, g, b)
                r, g, b = hsv2RGB(180, 100, v)
            elif param=='s':
                s = rgb2S(r, g, b)
                r, g, b = hsv2RGB(180, s, 100)
            draw.point((x,y), (r, g, b))
    return new_image
    

def biomLayer(img): #ДЛЯ ПЕРВОГО УРОВНЯ ГЕНЕРАЦИИ! (деление на биомы)
    width = img.size[0]
    height = img.size[1]
    print(width, height)
    new_image = img.resize((width,height), Image.ANTIALIAS)
    draw = ImageDraw.Draw(new_image)
    pix = new_image.load()
    for x in range(width):
        for y in range(height):
            r = pix[x,y][0]
            g = pix[x,y][1]
            b = pix[x,y][2]
            h = rgb2H(r, g, b)
            r, g, b = hsv2RGB(h, 100, 100)
            draw.point((x,y), (r, g, b))
    return new_image


def landscapeLayer(img): #ДЛЯ ВТОРОГО УРОВНЯ ГЕНЕРАЦИИ! (деление на поверхность)
    width = img.size[0]
    height = img.size[1]
    new_image = img.resize((width,height), Image.ANTIALIAS)
    draw = ImageDraw.Draw(new_image)
    pix = new_image.load()
    r_, g_, b_ = middleColor(img)
    mid = rgb2V(r_, g_, b_)
    for x in range(width):
        for y in range(height):
            r = pix[x,y][0]
            g = pix[x,y][1]
            b = pix[x,y][2]
            v = rgb2V(r, g, b)
            if v<mid:
                draw.point((x,y), (255, 0, 0))
            else:
                draw.point((x,y), (0, 0, 0))
    return new_image


def eventsLayer(img): #ДЛЯ ТРЕТЬЕГО УРОВНЯ ГЕНЕРАЦИИ! (деление на события)
    width = img.size[0]
    height = img.size[1]
    new_image = Image.new('RGB', (width,height), Image.ANTIALIAS)
    draw = ImageDraw.Draw(new_image)
    pix = img.load()
    r_, g_, b_ = middleColor(img)
    mid = rgb2S(r_, g_, b_)
    print(mid)
    for x in range(0, width, 2):
        for y in range(0, height, 2):
            r = pix[x,y][0]
            g = pix[x,y][1]
            b = pix[x,y][2]
            s = rgb2S(r, g, b)
            noise = (randint(-1, 1)) * (randint(-1, 1)) * randint(round(15*(r+g+b)/255), round(15*(r+g+b)/255)+10)
            shade = round(255-round(s*256/100)-mid)
            if randint(-3, 1) < 1: noise = 0
            elif randint(-2, 1) == 1:
                if shade >= mid: noise = -abs(noise)
                else: noise = abs(noise)
            draw.point((x,y), (shade+noise, shade+noise, shade+noise))
    return new_image

# !!!!!!!!!! СЮДА ВСТАВИТЬ НАЗВАНИЕ ФАЙЛА В РАСШИРЕНИЕМ
# !!!!!!!!!! ИЗ ТОЙ ЖЕ ДИРЕКТОРИИ  !!!!!!!
a = Image.open('test.png')

a = lowerImage(a)

img1 = biomLayer(a)
img2 = landscapeLayer(a)
img3 = eventsLayer(a)

num = '3'
img1.save(num+'testBioms.png', 'PNG')
img2.save(num+'testLandscape.png', 'PNG')
img3.save(num+'testEvents.png', 'PNG')

File: LevelGEN.cs
using UnityEngine;
using System.Collections.Generic;

/*
 * Класс LevelGEN - совокупность методов для генерации масок-слоёв уровня (твёрдая поверхность, биом),
 * а также деформация исходного изображения, работа с оттенками, вычисление среднего значения цвета по изображению
 */
public class LevelGEN
{
    Texture2D map;
    private List<int> subLayer;

    public void SetMap(Texture2D m)
    {
        map = m;
    }

    public float[] getRGB(int x, int y)
    {
        Color color = map.GetPixel(x, y);
        float[] colors = new float[3] { color.r, color.g, color.b };
        return colors;
    }

    public float rgb2H(float r, float g, float b)
    {
        r /= 255;
        g /= 255;
        b /= 255;
        float cmax = Mathf.Max(r, g, b);
        float cmin = Mathf.Min(r, g, b);
        float diff = cmax - cmin;

        float h = 0;
        if (cmax == cmin)
            h = 0;
        else if (cmax == r)
            h = (60 * ((g - b) / diff) + 360) % 360;
        else if (cmax == g)
            h = (60 * ((b - r) / diff) + 120) % 360;
        else if (cmax == b)
            h = (60 * ((r - g) / diff) + 240) % 360;
        return h;
    }

    public float rgb2S(float r, float g, float b)
    {
        r /= 255;
        g /= 255;
        b /= 255;
        float cmax = Mathf.Max(r, g, b);
        float cmin = Mathf.Min(r, g, b);
        float diff = cmax - cmin;
        float s = 0;
        if (cmax != 0)
            s = (diff / cmax) * 100;
        return s;
    }

    public float rgb2V(float r, float g, float b)
    {
        r /= 255;
        g /= 255;
        b /= 255;
        float cmax = Mathf.Max(r, g, b);
        float v = cmax * 100;
        return v;
    }

    public float[] hsv2RGB(float h, float s, float v)
    {
        float hi = Mathf.Floor(h / 60) % 6;
        float vm = (100 - s) * v / 100;
        float a = (v - vm) * (h % 60) / 60;
        float vi = vm + a;
        float vd = v - a;
        float[] colors = new float[3];
        if (hi == 0)
        {
            colors[0] = v;
            colors[1] = vi;
            colors[2] = vm;
        }
        else if (hi == 1)
        {
            colors[0] = vd;
            colors[1] = v;
            colors[2] = vm;
        }
        else if (hi == 2)
        {
            colors[0] = vm;
            colors[1] = v;
            colors[2] = vi;
        }
        else if (hi == 3)
        {
            colors[0] = vm;
            colors[1] = vd;
            colors[2] = v;
        }
        else if (hi == 4)
        {
            colors[0] = vi;
            colors[1] = vm;
            colors[2] = v;
        }
        else
        {
            colors[0] = v;
            colors[1] = vm;
            colors[2] = vd;
        }
        colors[0] = colors[0] * 255 / 100;
        colors[1] = colors[1] * 255 / 100;
        colors[2] = colors[2] * 255 / 100;
        return colors;
    }

    public float[] middleColor(Texture2D image)
    {
        float mid = 0; float r = 0; float g = 0; float b = 0;
        float[] colors;
        Color c;
        for (int x = 0; x < image.width; x++)
        {
            for (int y = 0; y < image.height; y++)
            {
                c = image.GetPixel(x, y);
                r += c.r;
                g += c.g;
                b += c.b;
                mid = mid + ((r + g + b) / 3);
            }
        }
        r /= (image.width * image.height);
        g /= (image.width * image.height);
        b /= (image.width * image.height);
        mid = mid / (image.width * image.height);
        colors = new float[4] { r, g, b, mid };
        return colors;
    }


    public Color[][] biomLayer(Texture2D image)
    {
        Color[][] layer = new Color[image.width][];
        for (int x = 0; x < image.width; x++)
        {
            layer[x] = new Color[image.height];
            for (int y = 0; y < image.height; y++)
            {
                float[] colors = getRGB(x, y);
                float h = rgb2H(colors[0], colors[1], colors[2]);
                float[] new_colors = hsv2RGB(h, 100, 100);
                layer[x][y].r = new_colors[0];
                layer[x][y].g = new_colors[1];
                layer[x][y].b = new_colors[2];
                layer[x][y].a = 255;
            }
        }
        return layer;
    }

    public float[][] landscapeLayer(Texture2D image)
    {
        subLayer = new List<int>();
        float[][] layer = new float[image.width][];
        float[] colors_ = middleColor(image);
        float mid = rgb2V(colors_[0], colors_[1], colors_[2]);
        for (int x = 0; x < image.width; x++)
        {
            layer[x] = new float[image.height];
            for (int y = 0; y < image.height; y++)
            {
                float[] colors = getRGB(x, y);
                float v = rgb2V(colors[0], colors[1], colors[2]);
                if (v < mid)
                {
                    layer[x][y] = 1;
                    subLayer.Add(x*10000+y);
                }
                else layer[x][y] = 0;
            }
        }
        return layer;
    }

    public List<float> goalsGenerate(int count)
    {
        List<float> layer = new List <float>();
        for (int i = 0; i < count; i++)
        {
            int ind = Random.Range(0, subLayer.Count - 1);
            layer.Add(subLayer[ind]);
        }
        return layer;
    }

    public Texture2D ScaleTexture(Texture2D source, int targetWidth, int targetHeight)
    {
        Texture2D result = new Texture2D(targetWidth, targetHeight, source.format, false);
        float incX = (1.0f / (float)targetWidth);
        float incY = (1.0f / (float)targetHeight);
        for (int i = 0; i < result.height; ++i)
        {
            for (int j = 0; j < result.width; ++j)
            {
                Color newColor = source.GetPixelBilinear((float)j / (float)result.width, (float)i / (float)result.height);
                result.SetPixel(j, i, newColor);
            }
        }
        result.Apply();
        return result;
    }
}
